src/vsr/replica.zig

```diff
@@ -7392,31 +7392,37 @@ pub fn ReplicaType(
             assert(op_max <= self.op);
 
             // Request enough prepares to utilize our max IO depth:
-            var budget = self.journal.writes.available();
-            if (budget == 0) {
+            var io_budget = self.journal.writes.available();
+            if (io_budget == 0) {
                 log.debug("{}: repair_prepares: waiting for IOP", .{self.replica});
                 return null;
             }
 
+            // Don't check the repair budget for a potential primary in view change, use the full
+            // write bandwidth for repair.
+            if (self.status == .normal and self.repair_messages_budget == 0) {
+                log.debug("{}: repair_prepares: waiting for repair budget", .{self.replica});
+                return null;
+            }
+
             var requested_op_max: ?u64 = null;
             for (op_min..op_max + 1) |op| {
                 if (self.journal.slot_with_op(op)) |slot| {
                     if (self.journal.dirty.bit(slot)) {
                         // Rebroadcast outstanding `request_prepare` every `repair_timeout` tick.
                         // Continue to request prepares until our budget is depleted.
-                        self.repair_messages_budget -|= 1;
-                        if (self.repair_messages_budget == 0) {
-                            log.debug("{}: repair_prepares: repair budget used", .{
-                                self.replica,
-                            });
-                            break;
-                        }
-
                         if (self.repair_prepare(op)) {
                             requested_op_max = op;
-                            budget -= 1;
-                            if (budget == 0) {
-                                log.debug("{}: repair_prepares: IO budget used", .{
+
+                            io_budget -= 1;
+                            if (io_budget == 0) {
+                                log.debug("{}: repair_prepares: IO budget used", .{self.replica});
+                                break;
+                            }
+
+                            self.repair_messages_budget -= @intFromBool(self.status == .normal);
```

Comment:
I think we need to decrement `repair_messages_budget` before we check whether `io_budget` is zero.